Wrap-up selectors package extraction

eliasdorneles opened the pull request:
Hey folks!

So, this is the result of the work in the sprints at Europython, @dangra, @kmike and @bbotella helped me to wrap up the selectors module extraction from Scrapy, and made sure it's working with Python 3.

We decided on the name **parsel** -- thanks @Granitas for the nice suggestion :+1: =)

Next things to do are update the docs and fix the docs build.

I was hoping @kmike and @dangra would give it a final review, and then after the docs update I believe we can release to PyPI.

What do you think?
Thanks everyone for the help, I had a great time doing this. :)

kmike commented on .travis.yml:
```diff
@@ -0,0 +1,16 @@
+  - "pypy"
+
+# command to install dependencies, e.g. pip install -r requirements.txt --use-mirrors
+install: pip install -r requirements.txt
```
What about running `python setup.py install` or `pip install .` instead? This way we'll be able to check missing dependencies in setup.py.

kmike commented on parsel/__init__.py:
```diff
@@ -0,0 +1,10 @@
+__email__ = 'info@scrapy.org'
+__version__ = '1.0.0'
+
+from parsel.unified import *
```
what about exposing the API explicitly? `from parsel.unified import Selector, ...`

kmike replied:
ah, I see, there is `__all__`

eliasdorneles replied:
+1, I'm in favor of replacing `__all__` for a explicit import.

kmike commented on parsel/unified.py:
```diff
@@ -32,22 +28,16 @@ def __init__(self, *args, **kwargs):
+class Selector(object):
 
     __slots__ = ['response', 'text', 'namespaces', 'type', '_expr', '_root',
                  '__weakref__', '_parser', '_csstranslator', '_tostring_method']
```
someof the slots are no longer needed

eliasdorneles replied:
I think it's only `response` that has got to go, right?
`__weakref__` is needed to support weak references -- we want that, right?

kmike replied:
yes, I think response is the only unneeded slot.

kmike commented on parsel/unified.py:
```diff
@@ -66,26 +56,28 @@ class Selector(object_ref):
-    def __init__(self, response=None, text=None, type=None, namespaces=None,
-                 _root=None, _expr=None):
-        self.type = st = _st(response, type or self._default_type)
+    def __init__(self, text=None, type=None, namespaces=None, _root=None, _expr=None):
```
what about renaming `root` argument - we're exposing it to user in `"Selector needs either text or _root argument"` exception message, so it is no longer private

eliasdorneles replied:
+1

kmike commented on parsel/unified.py:
```diff
@@ -66,26 +56,28 @@ class Selector(object_ref):
         self._expr = _expr
 
+    def _get_root(self, text):
+        body = text.strip().encode('utf8') or '<html/>'
```
if text is empty body is an unicode string, but if text is not empty body is a bytestring

eliasdorneles replied:
oops, nice catch!

eliasdorneles replied:
hmm, should we have an assertion for body not being bytes object?

kmike commented on setup.py:
```diff
@@ -0,0 +1,55 @@
+try:
+    from setuptools import setup
+except ImportError:
+    from distutils.core import setup
```
I think we can drop distutils support

eliasdorneles commented:
@kmike thanks for the review, I believe I've addressed all your comments.

@dangra I'm now just waiting for your review on how to proceed with the release. :)
Do you have some advice on how to configure Travis for it?
I'm tempted to just copy the stuff from here: https://github.com/scrapy/scrapy/blob/master/.travis.yml#L18 :smile:

kmike commented on parsel/unified.py:
```diff
@@ -66,26 +53,28 @@ class Selector(object_ref):
         self._expr = _expr
 
+    def _get_root(self, text):
+        body = text.strip().encode('utf8') or u'<html/>'
```
This doesn't look correct: if text is empty body becomes unicode (`u'<html/>'`), but if text is not empty body becomes a bytestring (`foo.encode('..')`).

AFAIK we require unicode here to unify data to utf8 before passing it to lxml. LXML can work on bytes (it'll try to detect the encoding), but we can do it better because there are other signs available like HTTP headers.

dangra commented:
I don't like makefiles, specially for running tests, the established way to run scrapy projects tests is tox. Running coverage and building docs looks convenient but still Make is a unix specific tool difficult to find in other systems.

Also about tests, I prefer if we stick to py.test and using TOXENV in travis-ci like we do for other scrapy projects: scrapy, queuelib and [w3lib travis.yml](https://github.com/scrapy/w3lib/blob/master/.travis.yml) serves as inspiration.

eliasdorneles commented:
@dangra alright, makes sense!
I'll get rid of the Makefile and move to py.test and tox.

eliasdorneles commented:
@dangra looks good to you?
I just copied the secure password from w3lib, would that be a problem?

dangra commented:
> I just copied the secure password from w3lib, would that be a problem?

not at all but it won't work because secrets are encrypted using per-repo secret keys.
